cirq/study/visualize.py

```diff
@@ -39,17 +37,18 @@ def plot_state_histogram(result: trial_result.TrialResult) -> np.ndarray:
     # This allows cirq to be usable without python3-tk.
     import matplotlib.pyplot as plt
 
-    num_qubits = len(result.measurements.keys())
+    assert set([value.shape[0] for value in result.measurements.values()]), \
+        'Same number of measurement must be made on all qubits.'
+    num_qubits = sum([value.shape[1] for value in result.measurements.values()])
+
     states = 2**num_qubits
     values = np.zeros(states)
-
     # measurements is a dict of {measurement gate key:
     #                            array(repetitions, boolean result)}
     # Convert this to an array of repetitions, each with an array of booleans.
     # e.g. {q1: array([[True, True]]), q2: array([[False, False]])}
     #      --> array([[True, False], [True, False]])
-    measurement_by_result = np.array([
-        v.transpose()[0] for k, v in result.measurements.items()]).transpose()
+    measurement_by_result = np.hstack(list(result.measurements.values()))
```

Comment:
Add more tests like `test_plot_state_histogram`.  One for each of the other two measurement cases discussed in the issue.

Reply:
Added the tests for this, for all the 3 cases. Applied `X` instead of `X ** 0.4` so that the results are deterministic and the tests are completed fast (can also do more repetitions and assert greater and less, if that would be better).